File: examples/legacy-architecture-samples/simple-stack-example-mvp-fragments/src/main/java/com/zhuinden/simplestackdemoexamplefragments/util/scopedservices/ServiceProvider.java
package com.zhuinden.simplestackdemoexamplefragments.util.scopedservices;

import androidx.annotation.NonNull;

import com.zhuinden.simplestack.ScopedServices;
import com.zhuinden.simplestack.ServiceBinder;

public class ServiceProvider implements ScopedServices {
    @Override
    public void bindServices(@NonNull ServiceBinder serviceBinder) {
        if(serviceBinder.getKey() instanceof HasServices) {
            ((HasServices) serviceBinder.getKey()).bindServices(serviceBinder);
        }
    }
}

File: tutorials/tutorial-sample/src/main/java/com/zhuinden/simplestacktutorials/steps/step_9/ServiceProvider.kt
package com.zhuinden.simplestacktutorials.steps.step_9

import com.zhuinden.simplestack.ScopedServices
import com.zhuinden.simplestack.ServiceBinder
import com.zhuinden.simplestacktutorials.steps.step_9.core.navigation.FragmentKey
import com.zhuinden.simplestacktutorials.steps.step_9.core.viewmodels.HasServices
import com.zhuinden.simplestacktutorials.steps.step_9.core.viewmodels.add
import com.zhuinden.simplestacktutorials.steps.step_9.features.registration.RegistrationViewModel

class ServiceProvider : ScopedServices {
    override fun bindServices(serviceBinder: ServiceBinder) {
        val key = serviceBinder.getKey<FragmentKey>()

        val scope = serviceBinder.scopeTag

        if (key is HasServices && key.scopeTag == scope) {
            key.bindServices(serviceBinder) // screen-bound shared services
        }

        with(serviceBinder) {
            when (scope) { // explicit shared services
                "registration" -> add(
                    RegistrationViewModel(
                        lookupService("appContext"),
                        backstack
                    )
                )
            }
        }
    }
}